tests/linearizability/identity/lease_ids.go

```diff
@@ -15,39 +14,151 @@
 package identity
 
 import (
+	"math/rand"
 	"sync"
 )
 
 type LeaseIdStorage interface {
-	LeaseId(int) int64
-	AddLeaseId(int, int64)
-	RemoveLeaseId(int)
+	// Add adds lease to storage.
+	Add(int64)
+
+	// Remove permanently removes lease from storage.
+	Remove(int64)
+
+	// Take returns an unused lease from storage
+	Take() int64
+
+	// Return returns a used lease to storage so it can be used later.
+	Return(int64)
+
+	// GetRandom returns a random lease (could be in use or not)
+	GetRandom() int64
+
+	// Count returns count of leases (could be in use or not)
+	Count() int
 }
 
 func NewLeaseIdStorage() LeaseIdStorage {
-	return &atomicClientId2LeaseIdMapper{m: map[int]int64{}}
+	return &atomicLeaseKeeper{ma: map[int64]int{}, mf: map[int64]int{}}
 }
 
-type atomicClientId2LeaseIdMapper struct {
+const noLease int64 = 0
+
+// atomicLeaseKeeper stores the leases in two lists - one for free leases and one for all leases.
+// two auxiliary maps are kept to allow fast deletion from the lists
+type atomicLeaseKeeper struct {
```

Comment:
This is very overdesigned. You just need a list with a mutex. No need to track in use leases, no need to do any mapping.

```
type LeaseStorage struct {
  mux sync.RWMutex
  free []int64
  inUse int
}

func (s *LeaseStorage) Add(leaseId int64) {
	s.mux.Lock()
	defer s.mux.Unlock()

	s.free = append(s.free, leaseId)
}

func (s *LeaseStorage) Remove(leaseId int64) {
	s.mux.Lock()
	defer s.mux.Unlock()
    s.remove(leaseId)
}

func (s *LeaseStorage) Take(toRemove int64) {
	s.mux.Lock()
	defer s.mux.Unlock()
	if s.remove(toRemove) {
      inUse++
    }
}

func (s *LeaseStorage) remove(toRemove int64) bool {
   i := 0 
   for _, lease := range s.free {
     if lease == toRemove {
       continue
     }
     s.free[i] = lease
     i++
   }
   return i != len(s.free)
}

func (s *LeaseStorage) Return(leaseId int64) {
	s.mux.Lock()
	defer s.mux.Unlock()
    inUse--
    s.free = append(s.free, leaseId)
}

func (s *LeaseStorage) Count() int {
	s.mux.RLock()
	defer s.mux.RUnlock()
    return len(s.free) + s.inUse
}
```